controllers/ocirepository_controller.go

```diff
@@ -0,0 +1,893 @@
+		}
+	}()
+	conditions.Delete(obj, sourcev1.StorageOperationFailedCondition)
+
+	var (
+		res      sreconcile.Result
+		resErr   error
+		metadata = sourcev1.Artifact{}
+	)
+
+	// Run the sub-reconcilers and build the result of reconciliation.
+	for _, rec := range reconcilers {
+		recResult, err := rec(ctx, obj, &metadata, tmpDir)
+		// Exit immediately on ResultRequeue.
+		if recResult == sreconcile.ResultRequeue {
+			return sreconcile.ResultRequeue, nil
+		}
+		// If an error is received, prioritize the returned results because an
+		// error also means immediate requeue.
+		if err != nil {
+			resErr = err
+			res = recResult
+			break
+		}
+		// Prioritize requeue request in the result.
+		res = sreconcile.LowestRequeuingResult(res, recResult)
+	}
+
+	r.notify(ctx, oldObj, obj, res, resErr)
+
+	return res, resErr
+}
+
+// reconcileSource fetches the upstream OCI artifact metadata and content.
+// If this fails, it records v1beta2.FetchFailedCondition=True on the object and returns early.
+func (r *OCIRepositoryReconciler) reconcileSource(ctx context.Context, obj *sourcev1.OCIRepository, metadata *sourcev1.Artifact, dir string) (sreconcile.Result, error) {
+	ctxTimeout, cancel := context.WithTimeout(ctx, obj.Spec.Timeout.Duration)
+	defer cancel()
+
+	options := r.craneOptions(ctxTimeout)
+
+	// Generate the registry credential keychain either from static credentials or using cloud OIDC
+	keychain, err := r.keychain(ctx, obj)
+	if err != nil {
+		e := serror.NewGeneric(
+			fmt.Errorf("failed to get credential: %w", err),
+			sourcev1.AuthenticationFailedReason,
+		)
+		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+	options = append(options, crane.WithAuthFromKeychain(keychain))
+
+	if obj.Spec.Provider != sourcev1.GenericOCIProvider {
+		auth, authErr := r.oidcAuth(ctxTimeout, obj)
+		if authErr != nil && !errors.Is(authErr, oci.ErrUnconfiguredProvider) {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to get credential from %s: %w", obj.Spec.Provider, authErr),
+				sourcev1.AuthenticationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+		if auth != nil {
+			options = append(options, crane.WithAuth(auth))
+		}
+	}
+
+	// Generate the transport for remote operations
+	transport, err := r.transport(ctx, obj)
+	if err != nil {
+		e := serror.NewGeneric(
+			fmt.Errorf("failed to generate transport for '%s': %w", obj.Spec.URL, err),
+			sourcev1.OCIOperationFailedReason,
+		)
+		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+	if transport != nil {
+		options = append(options, crane.WithTransport(transport))
+	}
+
+	// Determine which artifact revision to pull
+	url, err := r.getArtifactURL(obj, options)
+	if err != nil {
+		e := serror.NewGeneric(
```

Comment:
This can become `Stalling` error because if when URL is invalid, requeuing won't change anything. Need user intervention.
We do the same in GitRepositoryReconciler.

Reply:
No we can't, for semver it can match the range at a later time.

Reply:
`getArtifactURL()` is querying a remote server and also validating at the same time, which makes it hard to stall when we are sure something needs user intervention.
Maybe it would be better to return a typed error from `getArtifactURL()` such that `parseRepositoryURL()` error returns a stalling error and `getTagBySemver()` error returns a generic error with conditions set on the object accordingly.

Reply:
Done

Reply:
This implementation looks fine but I was more referring to something like the code around this https://github.com/fluxcd/source-controller/blob/b1a494993ab763b0ba6779b3485eb629310ef093/controllers/gitrepository_controller.go#L748-L752 , should have shared a reference before :sweat_smile: 
Don't think it's necessary to change. It's okay as it is.